### PR TITLE
New field resolver architecture proposal

### PR DESCRIPTION
**Pull request by rogyar**
### Description (*)
The general idea of the current PR is starting decoupling from Webonyx.
Currently, within a scope of a resolver, we work with data provided by Webonyx directly. For example:
https://github.com/magento/graphql-ce/blob/7b1cd79738b2cff7881a8defe4a58c6142c8ed88/app/code/Magento/QuoteGraphQl/Model/Resolver/Cart.php#L38-L43

In order to start the decoupling process we need to get rid of the direct usage of the Webonyx data and introduce an intermediate layer (interface) with the corresponding set of getters/setters. At the very beginning the mentioned intermediate layer will work with Webonyx. However, in the future we can change the interface implementation to our own if necessary.

### Fixed Issues (if relevant)
1. #71 : Decouple from Webonyx

### CODE REVIEW

File: lib/internal/Magento/Framework/GraphQl/FieldResolverContextInterface.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+declare(strict_types=1);
+
+namespace Magento\Framework\GraphQl;
+
+/**
+ * Interface FieldResolverContextInterface
+ *
+ * Field resolver context
+ */
+interface FieldResolverContextInterface
```
Comment (naydav) [inline lib/internal/Magento/Framework/GraphQl/FieldResolverContextInterface.php]: Please, rename to `QueryContextInterface`

File: lib/internal/Magento/Framework/GraphQl/Query/QueryInterface.php [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+declare(strict_types=1);
+
+namespace Magento\Framework\GraphQl\Query;
+
+/**
+ * Interface QueryInterface
+ *
+ * Contains information about GraphQl query
+ */
+interface QueryInterface
```
Comment (naydav) [inline lib/internal/Magento/Framework/GraphQl/Query/QueryInterface.php:15]: Please finish description this and the other of Interface (including parameters and return values)

File: lib/internal/Magento/Framework/GraphQl/Schema/Type/Output/ElementMapper/Formatter/Fields.php [hunk 3]
```diff
@@ -146,6 +146,10 @@ private function getFieldConfig(
             /** @var ResolverInterface $resolver */
             $resolver = $this->objectManager->get($field->getResolver());
 
+            // TODO: Here the new resolver interface will be substituted.
```
Comment (naydav) [inline lib/internal/Magento/Framework/GraphQl/Schema/Type/Output/ElementMapper/Formatter/Fields.php]: Please, implement the approach for any resolver from `QuoteGraphQl` module

File: lib/internal/Magento/Framework/GraphQl/Query/QueryInterface.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+{
+    /**
+     * TBP
+     */
+    public function getStructure();
+
+    /**
+     * TBP
+     */
+    public function getArgs();
```
Comment (naydav) [inline lib/internal/Magento/Framework/GraphQl/Query/QueryInterface.php]: Rename to `getArguments`

**Review by naydav — CHANGES_REQUESTED** [review]

**Comment by rogyar** [comment]
Hi @naydav. I've implemented Cart resolver according to the new resolver structure. 
I have doubts about the necessity to introduce a new context interface/implementation since we already have one `\Magento\Framework\GraphQl\Query\Resolver\ContextInterface`. I've reused this interface in my example. However, if you have another point on this matter - I'm open for discussions.

Thank you!

**Comment by naydav** [comment]
We are going to create a new more efficient way for filtration and loading data.
Creation of Design Document is in progress (#567).

So, need to wait to introduce a new mechanism and just after that continue work with this PR.

**Pull request closed without merging**